lit_gpt/model.py

```diff
@@ -289,6 +289,35 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.proj(x)
 
 
+class LLaMAMoE(nn.Module):
+    def __init__(self, config: Config) -> None:
+        super().__init__()
+        self.gate = nn.Linear(config.n_embd, config.n_expert, bias=False)
+        self.experts = nn.ModuleList(LLaMAMLP(config) for _ in range(config.n_expert))
+
+        self.config = config
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Derived from: https://github.com/dzhulgakov/llama-mistral/blob/cecee4/llama/model.py#L351-L364.
+        See also figure 1 in https://arxiv.org/abs/2211.15841
+        """
+        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
+        x = x.view(-1, C)  # (B*T, C)
+        router = self.gate(x)  # (B*T, n_expert)
+        probs, indices = torch.topk(router, self.config.n_expert_per_token, dim=-1)
+        probs = probs.softmax(dim=-1)  # (B*T, n_expert_per_token)
+        indices = indices.flatten()  # (B*T*n_expert_per_token)
+        x = x.repeat_interleave(self.config.n_expert_per_token, dim=0)
```

Comment:
Here is an interesting description from the [docs](https://pytorch.org/docs/stable/generated/torch.repeat_interleave.html#torch.repeat_interleave) of this method:

> Keyword Arguments
    output_size ([int](https://docs.python.org/3/library/functions.html#int), optional) – Total output size for the given axis ( e.g. sum of repeats). If given, it will avoid stream synchronization needed to calculate output shape of the tensor.

So, as I understand, if the `output_size` is not provided, in order to calculate it the stream synchronization will be called. Yikes.